velox/exec/GroupingSet.cpp

```diff
@@ -412,4 +446,244 @@ const HashLookup& GroupingSet::hashLookup() const {
   return *lookup_;
 }
 
+namespace {
+bool maybeReserve(int64_t increment, memory::MemoryUsageTracker& tracker) {
+  constexpr int32_t kGrowthQuantum = 8 << 20;
+  auto addedReservation = bits::roundUp(increment, kGrowthQuantum);
+  // We look up the tracker tree to see if there is a parent that could have
+  // space. If some parent could have space we take the chance and try to
+  // increase reservation.
+  auto candidate = &tracker;
+  while (candidate) {
+    auto limit = candidate->maxTotalBytes();
+    if (limit == memory::kMaxMemory && candidate->parent()) {
+      candidate = candidate->parent();
+      continue;
+    }
+    if (limit - candidate->getCurrentTotalBytes() > addedReservation) {
+      try {
+        tracker.reserve(addedReservation);
+      } catch (const std::exception& e) {
+        return false;
+      }
+      return true;
+    }
+    candidate = candidate->parent();
+  }
+  return false;
+}
+
+int64_t estimateSerializedSize(const VectorPtr& vector) {
+  vector_size_t bytes = 0;
+  auto bytesPtr = &bytes;
+  VectorStreamGroup::estimateSerializedSize(
+      vector,
+      std::vector<IndexRange>{IndexRange{0, vector->size()}},
+      &bytesPtr);
+  return bytes;
+}
+} // namespace
+
+void GroupingSet::ensureInputFits(const RowVectorPtr& input) {
+  if (isPartial_ || spillPath_.empty()) {
+    return;
+  }
+  auto numDistinct = table_->numDistinct();
+  if (!numDistinct) {
+    return;
+  }
+  auto tableIncrement = numDistinct + input->size() >= table_->rehashSize()
+      ? BaseHashTable::tableByteSize(2 * numDistinct) -
+          BaseHashTable::tableByteSize(numDistinct)
+      : 0;
+
+  int64_t retainedSize = input->retainedSize();
+
+  auto rows = table_->rows();
+
+  auto [freeRows, outOfLineFreeBytes] = rows->freeSpace();
+  auto outOfLineBytes =
+      rows->stringAllocator().retainedSize() - outOfLineFreeBytes;
+  auto outOfLineBytesPerRow =
+      (outOfLineBytes / numDistinct) + minVariableWidthAccumulatorBytes_;
```

Comment:
Instead of relying on individual aggregates to provide reasonable estimates, can we use rows->stringAllocator().retainedSize() / numDistinct as the estimate?